### PR TITLE
Refactor/scheduler job listing

### PR DESCRIPTION
**Pull request by belerophon**
This PR changes the scheduler interface, and makes the methods `scheduleAction` and `getJobsStatus` return the scheduler action and a map of ids to actions, respectively. The first method, returns the action that was just scheduled, if no error occurs. The second one, gives us a map of all the currently scheduled actions, indexed by action id.

This interface seems to be more useful than the one where only a list of ids were returned. Also, the amount of data to return for each of the available scheduler actions does not seem to prohibit this approach.

### CODE REVIEW

File: src/main/scala/eu/shiftforward/adstax/scheduler/action/SchedulerAction.scala [hunk 1]
```diff
@@ -51,19 +52,15 @@ object SchedulerAction {
...
       }
 
       def write(a: SchedulerAction): JsValue = a match {
-        case a: SchedulerAction if jsonFormatRegistry.contains(a.action) =>
-          a.toJson(jsonFormatRegistry(a.action))
-        case other =>
-          throw new SerializationException(s"Cannot serialize unrecognized SchedulerAction: ${other.getClass}!")
+        case a: SchedulerAction if jsonFormatRegistry.contains(a.action) => a.toJson(jsonFormatRegistry(a.action))
+        case other => serializationError(s"Cannot serialize unrecognized Scheduler Action: ${other.getClass}!" +
+          s" Does the JSON Format Registry contain a '${a.action}' entry for that Scheduler Action!?")
```
Comment (luismfonseca) [inline src/main/scala/eu/shiftforward/adstax/scheduler/action/SchedulerAction.scala]: +1 for thoughtfulness

File: src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerClient.scala [hunk 2]
```diff
@@ -13,7 +14,7 @@ trait SchedulerClient {
    * @param action the action to be scheduled
    * @return a `Future` with the id of the scheduled action.
```
Comment (jcazevedo) [inline src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerClient.scala]: Update the ScalaDoc here. We're no longer returning the id.

File: src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerClient.scala [hunk 3]
```diff
@@ -28,7 +29,7 @@ trait SchedulerClient {
    *
    * @return a `Future` with the list of all previously scheduled job status.
```
Comment (jcazevedo) [inline src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerClient.scala]: Update the ScalaDoc here. We're no longer returning the the list of all previously scheduled job status.

File: src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerOperationResult.scala [hunk 4]
```diff
@@ -1,38 +1,37 @@
...
- * `SchedulerActor`.
- *
- * The reply can be a successful one, or an error
- */
-sealed trait SchedulerStatus {
-  /**
-   * The id of the `SchedulerAction`.
-   */
-  def id: String
+  * The reply type for all Schedule requests
```
Comment (jcazevedo) [inline src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerOperationResult.scala]: We tend to align the `*` in lines after the first with the first `*` in the first line. For consistency, can you update the comments on this file? I think it is the only one in this PR with a different styling.

File: src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerOperationResult.scala [hunk 5]
```diff
@@ -66,80 +65,79 @@ object SchedulerOperationResult {
...
 
-    // reads and writes SchedulerStatus
-    implicit object SchedulerStatusJsonFormat extends RootJsonFormat[SchedulerStatus] {
+    implicit object ScheduledReplyJsonFormat extends RootJsonFormat[ScheduledReply] {
       private val scheduledFormat = jsonFormat1(Scheduled)
       private val scheduledErrorFormat = jsonFormat2(ScheduledError)
+
+      override def read(json: JsValue): ScheduledReply = json.asJsObject.getFields("status") match {
+        case Seq(JsString("scheduled")) => scheduledFormat.read(json).asInstanceOf[ScheduledReply]
+        case Seq(JsString("scheduled-error")) => scheduledErrorFormat.read(json).asInstanceOf[ScheduledReply]
```
Comment (jcazevedo) [inline src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerOperationResult.scala]: Are these `asInstanceOf[ScheduledReply]` necessary?

File: src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerOperationResult.scala [hunk 6]
```diff
@@ -66,80 +65,79 @@ object SchedulerOperationResult {
...
+    implicit object CancelledReplyJsonFormat extends RootJsonFormat[CancelledReply] {
       private val cancelledFormat = jsonFormat1(Cancelled)
       private val cancelledErrorFormat = jsonFormat2(CancelledError)
 
-      override def read(json: JsValue): SchedulerStatus = json.asJsObject.getFields("status") match {
-        case Seq(JsString("scheduled")) => scheduledFormat.read(json).asInstanceOf[SchedulerStatus]
-        case Seq(JsString("scheduled-error")) => scheduledErrorFormat.read(json).asInstanceOf[SchedulerStatus]
+      override def read(json: JsValue): CancelledReply = json.asJsObject.getFields("status") match {
         case Seq(JsString("cancelled")) => cancelledFormat.read(json).asInstanceOf[CancelledReply]
         case Seq(JsString("cancelled-error")) => cancelledErrorFormat.read(json).asInstanceOf[CancelledReply]
```
Comment (jcazevedo) [inline src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerOperationResult.scala]: Are these `asInstanceOf[CancelledReply]` necessary?

File: src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerOperationResult.scala [hunk 7]
```diff
@@ -66,80 +65,79 @@ object SchedulerOperationResult {
...
+      override def read(json: JsValue): CancelledError = CancelledReplyJsonFormat.read(json).asInstanceOf[CancelledError]
+      override def write(obj: CancelledError): JsValue = CancelledReplyJsonFormat.write(obj)
     }
 
     // reads and writes SchedulingResponseMessage
     implicit object SchedulingResponseMessageJsonFormat extends RootJsonFormat[SchedulingResponseMessage] {
-      private val jobsStatusFormat: RootJsonFormat[JobsStatus] = new RootJsonFormat[JobsStatus] {
-        override def write(js: JobsStatus) = JsObject("jobs" -> js.jobs.map(_.toJson).toJson)
 
+      implicit private val jobsStatusFormat: RootJsonFormat[JobsStatus] = new RootJsonFormat[JobsStatus] {
```
Comment (jcazevedo) [inline src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerOperationResult.scala]: There's a `RootJsonFormat[JobsStatus]` already available in the `JsonProtocol` object. Its implementation differs from this one, though. Is it intended for the `SchedulingResponseMessage` to have a different format for `JobsStatus`? If so, the tests should reflect that (I replaced this format with the one available in the `JsonProtocol` object and the tests pass).

Comment (belerophon) [inline src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerOperationResult.scala]: Like we've spoken, this is the private format. The other one is just a wrapper for this one. I've added some tests for the rest of the messages on c863676

**Comment by belerophon** [comment]
@jcazevedo I think I've addressed your comments on eb60454, 6b92ea6 and c863676.
Please check them out.

File: src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerOperationResult.scala (hunk 7 above)
Comment (jcazevedo) [inline src/main/scala/eu/shiftforward/adstax/scheduler/SchedulerOperationResult.scala]: This still felt a bit weird to me, and we were doing some unnecessary hops to serialize/deserialize certain types. I made some changes in db65f29. If you agree with them, you can merge this in.

**Pull request merged**